core/src/main/java/org/apache/calcite/util/RuleMatchVisualizerHook.java

```diff
@@ -0,0 +1,155 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.util;
+
+import org.apache.calcite.config.CalciteConnectionConfig;
+import org.apache.calcite.jdbc.CalciteConnection;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.hep.HepPlanner;
+import org.apache.calcite.plan.visualizer.RuleMatchVisualizer;
+import org.apache.calcite.runtime.Hook;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+/**
+ * Utility class to enable RuleMatchVisualizer for Calcite connections.
+ *
+ * <p>This class provides hooks to automatically attach a RuleMatchVisualizer
+ * to planners when a connection specifies the ruleVisualizerDir property.
+ *
+ * <p>Usage in JDBC URL:
+ * <blockquote><pre>
+ * jdbc:calcite:ruleVisualizerDir=/tmp/calcite-viz
+ * </pre></blockquote>
+ *
+ * <p>Or programmatically:
+ * <blockquote><pre>
+ * RuleMatchVisualizerHook.enable("/tmp/calcite-viz");
+ * </pre></blockquote>
+ */
+public class RuleMatchVisualizerHook {
+  public static final RuleMatchVisualizerHook INSTANCE = new RuleMatchVisualizerHook();
+
+  private final Map<RelOptPlanner, RuleMatchVisualizer> visualizerMap = new HashMap<>();
+  private final AtomicInteger queryCounter = new AtomicInteger(0);
+
+  private Hook.Closeable hookCloseable = Hook.Closeable.EMPTY;
+
+  /** Private constructor to prevent instantiation. */
+  private RuleMatchVisualizerHook() {}
+
+  /**
+   * Enables the visualizer for all subsequent queries with the specified output directory.
+   *
+   * @param outputDir Directory where visualization files will be created
+   */
+  public synchronized void enable(String outputDir) {
+    hookCloseable.close();
+
+    // Ensure the output directory exists
+    File dir = new File(outputDir);
+    if (!dir.exists()) {
+      boolean madeDir = dir.mkdirs();
+      assert madeDir : "Failed to create directory: " + outputDir;
+    }
+
+    // Install the hook
+    hookCloseable = Hook.PLANNER.addThread((Consumer<RelOptPlanner>) planner -> {
+      attachVisualizer(planner, outputDir);
+    });
+  }
+
+  /**
+   * Enables the visualizer using the connection's configuration.
+   * This method checks if the connection has the ruleVisualizerDir property set.
+   *
+   * @param connection The Calcite connection
+   */
+  public synchronized void enableFromConnection(CalciteConnection connection) {
+    CalciteConnectionConfig config = connection.config();
+    String vizDir = config.ruleVisualizerDir();
+
+    if (vizDir != null && !vizDir.isEmpty()) {
+      enable(vizDir);
+    }
+  }
+
+  /**
+   * Disables the visualizer.
+   */
+  public synchronized void disable() {
```

Comment:
The `disable()` should be called somewhere as well. The Handler defined in `org.apache.calcite.jdbc.Driver#createHandler` could override `onConnectionClose` to call it. I would strongly recommend to make the calls to "enable" and "disable" in the same class. So either moving the init code to `Driver#createHandler` as well, or creating another method in `CalciteConnectionImpl` to complement the `init()` method.